### PR TITLE
BUILD-401: add an openshift build using entitlements via shared resources example

### PR DESCRIPTION
**Pull request by gabemontero**
/assign @coreydaley 
for general code review
/assign @jitendar-singh 
/assign @prietyc123 
for qe approved label - the doc here are my instructions on how to manually verify this scenario
/assign @rolfedh 
for docs approved label - this doc here is my input for the "connecting of the dots" item that @siamaksade opened against you for explaining how to use insights operator, shared resource operator, and openshift builds to do entitled buidls *WITHOUT* copying the subscription credentials into a secret in the same namespace as the build config, as we currently document as the way to do entitlements with openshift builds in https://docs.openshift.com/container-platform/4.9/cicd/builds/running-entitled-builds.html

@openshift/openshift-team-build-api-maintainers FYI

As I understood @adambkaplan during planning today, the next story where we actually create the blog post is where we "officially" engage with PX and get their comfort level / reqs on all this, so 

/label px-approved

### CODE REVIEW

File: docs/simple-example.md [hunk 1]
```diff
@@ -159,11 +87,46 @@ How `Secrets` and `ConfigMaps` are stored on disk mirror the storage for
...
+$ oc start-build my-csi-bc -F
+```
+
+and what as Red Hat entitled content via your subscription is installed during the running of the Docker Strategy build.
+See the next section first however, if you have never used OpenShift Insights Operator to inject your subscription 
+credentials into your cluster.
+
+## Notes or setting up your cluster so that Insights Operator injects your subscription credentials for use in your cluster
+
+So both OpenShift Shared Resources and OpenShift Insights Operator are only available on Tech Preview clusters.
```
Comment (adambkaplan) [inline docs/simple-example.md:104]: Starting a sentence with "So" is very informal, and isn't necessary. In addition, the Insights Operator itself is available on all clusters - only the "download Simple Content Accesss cert" is tech preview.

```suggestion
Tech Preview must be enabled on your OCP cluster to use the Shared Resources CSI Driver and to sync the Simple Content Access certificate of your Red Hat account.
```

File: docs/simple-example.md [hunk 2]
```diff
@@ -159,11 +87,46 @@ How `Secrets` and `ConfigMaps` are stored on disk mirror the storage for
 where a file is created for each key in a `ConfigMap` `data` map or `binaryData` map and each key in a `Secret`
 `data` map.
 
-If you want to try other `ConfigMaps` or a `Secret`, first clear out the existing application:
+## BuildConfig example
```
Comment (adambkaplan) [inline docs/simple-example.md:90]: This section is sufficiently complex that I would put it in a separate document.

**Review by adambkaplan — CHANGES_REQUESTED** [review]

**Comment by adambkaplan** [comment]
@gabemontero a note regarding PR descriptions - both the PR title and description are used to generate customer-facing changelogs. Therefore, please update the PR title and description so that it is readable to an end-user.

Team member mentions should be placed in separate comments.

File: docs/simple-example.md [hunk 3]
```diff
@@ -1,102 +1,30 @@
...
 needed to illustrate the mounting of one of the API types (in this instance a `ConfigMap` from the `openshift-config`
-namespace) into the `Pod`
+namespace) into the `Pod`.
 
-For references, some in-lined yaml as a starting point.  Feel free to adjust the details to a working alternative that
-meets your needs:
+The second form is an OpenShift `BuildConfig` that utilizes a `SharedSecret` of the `etc-pki-entitlement` secret in the
+`openshift-config-managed` namespace, which is set up by the OpenShift Insights Operator (which is currently a Tech Preview
+only feature for OCP just like the Shared Resources CSI Driver) if your install pull secret in fact has a valid 
+Red Hat subscription that entitles you to install Red Hat RPMs.  With a referenced to this `SharedSecret`, the `BuildConfig`
```
Comment (coreydaley) [inline docs/simple-example.md:12]:
```suggestion
Red Hat subscription that entitles you to install Red Hat RPMs.  With a reference to this `SharedSecret`, the `BuildConfig`
```

File: docs/simple-example.md [hunk 4]
```diff
@@ -1,102 +1,30 @@
...
+is able to use your subscription information to installed entitled RPMs when building an image.  Separate `Role` and
```
Comment (coreydaley) [inline docs/simple-example.md:13]:
```suggestion
is able to use your subscription information to installe entitled RPMs when building an image.  Separate `Role` and 
```

File: docs/simple-example.md [hunk 5]
```diff
@@ -1,102 +1,30 @@
...
+namespace) into the `Pod`.
 
-For references, some in-lined yaml as a starting point.  Feel free to adjust the details to a working alternative that
-meets your needs:
+The second form is an OpenShift `BuildConfig` that utilizes a `SharedSecret` of the `etc-pki-entitlement` secret in the
+`openshift-config-managed` namespace, which is set up by the OpenShift Insights Operator (which is currently a Tech Preview
+only feature for OCP just like the Shared Resources CSI Driver) if your install pull secret in fact has a valid 
+Red Hat subscription that entitles you to install Red Hat RPMs.  With a referenced to this `SharedSecret`, the `BuildConfig`
+is able to use your subscription information to installed entitled RPMs when building an image.  Separate `Role` and 
+`RoleBinding` objects are defined for the `BuildConfig` and its use of the `SharedSecret`
```
Comment (coreydaley) [inline docs/simple-example.md:14]:
```suggestion
`RoleBinding` objects are defined for the `BuildConfig` and its use of the `SharedSecret`.
```

File: docs/simple-example.md [hunk 6]
```diff
@@ -159,11 +87,46 @@ How `Secrets` and `ConfigMaps` are stored on disk mirror the storage for
...
+
+You can simply start a `Build` for this `BuildConfig`, and follow the logs, via 
 
 ```shell
-$ oc delete -f ./examples 
-``` 
+$ oc start-build my-csi-bc -F
+```
+
+and what as Red Hat entitled content via your subscription is installed during the running of the Docker Strategy build.
```
Comment (coreydaley) [inline docs/simple-example.md:98]:
```suggestion
and watch as Red Hat entitled content via your subscription is installed during the running of the Docker Strategy build.
```

**Review by coreydaley — CHANGES_REQUESTED** [review]

**Comment by gabemontero** [comment]
updates made @adambkaplan PTAL thanks

**Comment by gabemontero** [comment]
I'll have to do your suggested commits manually @coreydaley since I moved that content to a new file per @adambkaplan but I'll get that started soon thanks

**Comment by gabemontero** [comment]
@adambkaplan hold a bit on reviewing for you comments ... @coreydaley 's comments made me realize I need to move some doc content from the old file to the new one

**Comment by gabemontero** [comment]
ok @adambkaplan @coreydaley I think I have both your sets of comments reconciled and applied together ... PTAL when you get the chance - thanks

**Comment by gabemontero** [comment]
bump @rolfedh 

for doc approve label ... the new markdowns serve as my input to you for https://issues.redhat.com/browse/RHDEVDOCS-3631

**Comment by gabemontero** [comment]
bump @jitendar-singh @prietyc123 

for qe approve label

this pr provides the yaml and instructions for you to try entitled builds yourself as part of manual verification (along with an associated feature file in the `smoke` dir for https://issues.redhat.com/browse/BUILD-400

**Comment by gabemontero** [comment]
@adambkaplan @jitendar-singh @prietyc123 how do you all want to handle the assignment of https://issues.redhat.com/browse/BUILD-400 ? ... it is currently unassigned

**Comment by gabemontero** [comment]
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_csi-driver-shared-resource/83/pull-ci-openshift-csi-driver-shared-resource-master-e2e-aws-csi-driver-disruptive/1481256911906017280 was an AWS fail

File: README.md [hunk 7]
```diff
@@ -49,8 +49,11 @@ or namespace administrator.
 
 The full definition of the `SharedConfigMap` can be found [here](deploy/0000_10_sharedconfigmap.crd.yaml) or `SharedSecret` custom resource can be found [here](deploy/0000_10_sharedsecret.crd.yaml).
 
-For a more real world example of using this new driver to help with sharing RHEL entitlements, [this blog post](https://www.openshift.com/blog/the-path-to-improving-the-experience-with-rhel-entitlements-on-openshift)
-dives into that scenario.
+Under the [examples directory](examples) there is both a simple example of using a `SharedConfigMap` in a `Pod`, and an
+example of performing an OpenShift `Build` from a `BuildConfig` that uses Red Hat Entitlements available from your Red Hat Subscription when
+your subscription credentials are stored on your cluster via OpenShift Insights Operator, and resulting `Secret` for those
+credentials is made available to additional `Namespaces` via a `SharedSecret`.  Instructions for the `Pod` example are
+[here](docs/simple-example.md), and for the `BuildConfig` example are [here]().
```
Comment (coreydaley) [inline README.md:56]: missing link

**Comment by gabemontero** [comment]
link added thanks @coreydaley 

ptal @coreydaley @adambkaplan

**Comment by gabemontero** [comment]
aws flake on https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_csi-driver-shared-resource/83/pull-ci-openshift-csi-driver-shared-resource-master-e2e-aws-csi-driver-slow/1481360209472393216

/retest

**Comment by gabemontero** [comment]
no more flakes - all tests are green

@adambkaplan @coreydaley PTAL

**Comment by gabemontero** [comment]
@rolfedh - I've added a section to the google doc input into the new "Shared Resources" docs that is a starting point for its "connect the docs" piece ...i.e. connecting the dots between the doc articles around Shared Resources, Builds, and Insight Operator/Simple Content Access.

**Comment by gabemontero** [comment]
ok @coreydaley rebase handled (it was simple, an existing file was renamed in this PR)

PTAL / retag for merge

**Comment by adambkaplan** [comment]
/assign @prietyc123 @jitendar-singh 

for qe approval.

**Pull request merged**